### PR TITLE
allow calling `goToDef` on modifiers of named declarations

### LINKED ISSUE #60308: Allow running `go to definition` on `export` and other keywords
### 🔎 Search Terms

- go to definition

### 🕗 Version & Regression Information

Not a regression

### ⏯ Playground Link

_No response_

### 💻 Code

```ts
export class A {

    private z: string;

    readonly x: string;

    async a() {  }

    override b() {}
}
```

Try running `go to definition` or `go to type definition` on any of the keywords before the symbol name (`export`, `private`, `override`, `async`)

### 🙁 Actual behavior

No definitions fourd

### 🙂 Expected behavior

It should work the same as running go to definition on the `symbol` name

### Additional information about the issue

_No response_

### PR DESCRIPTION
**Pull request by iisaduan**
Fixes #60308.

There are some symbols where goToDef/goToType previously did not return a definition. For those cases, I have not changed the behavior, this PR just makes it so that calling `goTo(Type)Definition` on a modifier will do the same behavior as if the call was triggered on the name.

The first commit refactors some existing `goToDefinition` keyword cases. It may be easier to review by looking at the first commit, and then the rest of the commits without the first one.

### CODE REVIEW

**Review by mjbvz — APPROVED** [review]
Thanks for looking into this!

**Review by sandersn — APPROVED** [review]

File: src/services/goToDefinition.ts [hunk 1]
```diff
@@ -159,16 +157,30 @@ export function getDefinitionAtPosition(program: Program, sourceFile: SourceFile
...
-        const functionDeclaration = findAncestor(node, n => isFunctionLikeDeclaration(n));
-        const isAsyncFunction = functionDeclaration && some(functionDeclaration.modifiers, node => node.kind === SyntaxKind.AsyncKeyword);
-        return isAsyncFunction ? [createDefinitionFromSignatureDeclaration(typeChecker, functionDeclaration)] : undefined;
+    // for keywords related to function or method definitions
+    let findFunctionDecl: ((n: Node) => boolean | "quit") | undefined;
+    let checkFunctionDeclaration: ((decl: FunctionLikeDeclaration) => boolean) | undefined;
+    switch (node.kind) {
+        case SyntaxKind.ReturnKeyword:
+            findFunctionDecl = n => {
+                return isClassStaticBlockDeclaration(n)
```
Comment (DanielRosenwasser) [inline src/services/goToDefinition.ts:166]: This feels like it's very much a special-case. Are you trying to just not break some sort of test case?

File: src/services/goToDefinition.ts [hunk 2]
```diff
@@ -159,16 +157,30 @@ export function getDefinitionAtPosition(program: Program, sourceFile: SourceFile
...
     }
-
-    if (node.kind === SyntaxKind.YieldKeyword) {
-        const functionDeclaration = findAncestor(node, n => isFunctionLikeDeclaration(n));
-        const isGeneratorFunction = functionDeclaration && functionDeclaration.asteriskToken;
-        return isGeneratorFunction ? [createDefinitionFromSignatureDeclaration(typeChecker, functionDeclaration)] : undefined;
+    if (findFunctionDecl) {
+        const functionDeclaration = findAncestor(node, findFunctionDecl) as FunctionLikeDeclaration | undefined;
+        const isCorrectDeclaration = functionDeclaration && (!checkFunctionDeclaration || checkFunctionDeclaration(functionDeclaration));
+        return isCorrectDeclaration ? [createDefinitionFromSignatureDeclaration(typeChecker, functionDeclaration)] : undefined;
```
Comment (DanielRosenwasser) [inline src/services/goToDefinition.ts:183]: I kind of find the indirection here hard to follow. I can see wanting to find some common logic, but I'd suggest just writing it out explicitly in each branch for now.

File: src/services/goToDefinition.ts (hunk 1 above)
Comment (iisaduan) [reply src/services/goToDefinition.ts:166]: It was existing code, so I'm not sure if there are more cases that would be an issue, but it breaks https://github.com/microsoft/TypeScript/blob/main/tests/cases/fourslash/goToDefinitionReturn5.ts

File: src/services/goToDefinition.ts (hunk 2 above)
Comment (DanielRosenwasser) [reply src/services/goToDefinition.ts:183]: Also, the idea that you *wouldn't* want to still jump to the nearest function even if it's not `async` or a generator is something I'm not so sure of. So the idea that you need a `checkFunctionDeclaration` to validate that you got a function of the right kind feels like it would break down unnecessarily during refactoring.

**Review by DanielRosenwasser — DISMISSED** [review]

File: src/services/goToDefinition.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/services/goToDefinition.ts:166]: It's probably just fine to update the baseline there.

**Comment by iisaduan** [comment]
@mjbvz I addressed @DanielRosenwasser 's comments and added some more keywords to return definitions. Would you both take a look to see if the new behavior is reasonable?

**Review by mjbvz — APPROVED** [review]
Thanks @iisaduan! Just tested this in VS Code and the new behavior looks great. Thank you for looking into this

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this
@typescript-bot test tsserver top300

**Review by RyanCavanaugh — APPROVED** [review]
